### PR TITLE
Optimize location hierarchy

### LINKED ISSUE #82: FHIR Gateway Performance Optimization
We need to explore further optimizations for the FHIR Gateway especially for the Related Entity Location Sync Strategy

### PR DESCRIPTION
**Pull request by lincmba**
-  Instead of fetching entire practitionerdetails then getting locations, just get the locations.
- replace for loops with streams
 - use sets instead of lists to avoid duplication
 - switch from using DFS to BFS in getting locations deep in the hierarchy

Resolves #82 

**Engineer Checklist**

- [ ] I have written **Unit tests** for any new feature(s) and edge cases for
      bug fixes
- [ ] I have added documentation for any new feature(s) and configuration
      option(s) on the `README.md`
- [ ] I have run `mvn spotless:check` to check my code follows the project's
      style guide
- [ ] I have run `mvn clean test jacoco:report` to confirm the coverage report
      was generated at `plugins/target/site/jacoco/index.html`
- [ ] I ran `mvn clean package` right before creating this pull request.

### CODE REVIEW

File: plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java [hunk 1]
```diff
@@ -443,23 +453,25 @@ public List<String> generateAdminLevels(
 
     public List<Location> filterLocationsByAdminLevels(
             List<Location> locations, List<String> postFetchAdminLevels) {
-        if (postFetchAdminLevels == null) {
+
+        if (postFetchAdminLevels == null || postFetchAdminLevels.isEmpty()) {
```
Comment (lincmba) [inline plugins/src/main/java/org/smartregister/fhir/gateway/plugins/LocationHierarchyEndpointHelper.java:457]: Resolves bug of filtering out all locations when adminlevels is not provided/is empty

**Comment by ndegwamartin** [comment]
PR looks good. Lets bump up the Release version as well cc @lincmba

**Review by ndegwamartin — APPROVED** [review]
